src/Catalog/Icons/IconCopyResultCache.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NuGet.Services.Metadata.Catalog.Persistence;
+
+namespace NuGet.Services.Metadata.Catalog.Icons
+{
+    public class IconCopyResultCache : IIconCopyResultCache, IIconCopyResultCachePersistence
+    {
+        private const string CacheFilename = "c2i_cache.json";
+
+        private static ConcurrentDictionary<Uri, ExternalIconCopyResult> ExternalIconCopyResults = null;
+
+        private readonly IStorage _auxStorage;
+
+        public IconCopyResultCache(
+            IStorage auxStorage)
+        {
+            _auxStorage = auxStorage ?? throw new ArgumentNullException(nameof(auxStorage));
+        }
+
+        public async Task InitializeIconUrlCacheAsync(CancellationToken cancellationToken)
```

Comment:
Generally, the method names here can be simplified: Initialize, SaveAsync, Get, Set